be/dtm0_log.c

```diff
@@ -0,0 +1,658 @@
+		m0_free(lrec);
+		return rc;
+	}
+
+	rc = m0_buf_copy(&lrec->dlr_payload, payload);
+	if (rc != 0) {
+		m0_dtm0_tx_desc_fini(&lrec->dlr_txd);
+		m0_free(lrec);
+		return rc;
+	}
+
+	*rec = lrec;
+	return 0;
+}
+
+static int be_dtm0_plog_rec_init(struct m0_dtm0_log_rec **out,
+				 struct m0_be_tx         *tx,
+				 struct m0_be_seg        *seg,
+				 struct m0_dtm0_tx_desc  *txd,
+				 struct m0_buf           *payload)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	M0_BE_ALLOC_PTR_SYNC(rec, seg, tx);
+	M0_ASSERT(rec != NULL);
+
+	rec->dlr_txd.dtd_id = txd->dtd_id;
+	rec->dlr_txd.dtd_ps.dtp_nr = txd->dtd_ps.dtp_nr;
+	M0_BE_ALLOC_ARR_SYNC(rec->dlr_txd.dtd_ps.dtp_pa,
+			     txd->dtd_ps.dtp_nr, seg, tx);
+	M0_ASSERT(rec->dlr_txd.dtd_ps.dtp_pa != NULL);
+
+	memcpy(rec->dlr_txd.dtd_ps.dtp_pa, txd->dtd_ps.dtp_pa,
+	       sizeof(rec->dlr_txd.dtd_ps.dtp_pa[0]) * rec->dlr_txd.dtd_ps.dtp_nr);
+
+	if (payload->b_nob > 0) {
+		rec->dlr_payload.b_nob = payload->b_nob;
+		M0_BE_ALLOC_BUF_SYNC(&rec->dlr_payload, seg, tx);
+		M0_ASSERT(&rec->dlr_payload.b_addr != NULL); /* TODO: handle error */
+		m0_buf_memcpy(&rec->dlr_payload, payload);
+	} else {
+		rec->dlr_payload.b_addr = NULL;
+		rec->dlr_payload.b_nob = 0;
+	}
+
+	M0_BE_TX_CAPTURE_BUF(seg, tx, &rec->dlr_payload);
+	M0_BE_TX_CAPTURE_ARR(seg, tx,
+			     rec->dlr_txd.dtd_ps.dtp_pa,
+			     rec->dlr_txd.dtd_ps.dtp_nr);
+	M0_BE_TX_CAPTURE_PTR(seg, tx, rec);
+
+	*out = rec;
+	return 0;
+}
+
+/* TODO: change convention of this function to:
+void be_dtm0_log_rec_fini(struct m0_dtm0_log_rec *rec, ..*tx)
+and allocate rec outside the function
+*/
+static void be_dtm0_log_rec_fini(struct m0_dtm0_log_rec **rec,
+				 struct m0_be_tx         *tx)
+{
+	struct m0_dtm0_log_rec *lrec = *rec;
+
+	M0_ASSERT_INFO(M0_IS0(&lrec->dlr_dtx), "DTX should have been finalised "
+		       "already in m0_dtx0_done().");
+	m0_buf_free(&lrec->dlr_payload);
+	m0_dtm0_tx_desc_fini(&lrec->dlr_txd);
+	m0_free(lrec);
+	*rec = NULL;
+}
+
+/* TODO: change convention of this function to:
+void m0_be_dtm0_plog_rec_fini(struct m0_dtm0_log_rec *rec, ..*tx)
+and allocate rec outside the function
+*/
+static void be_dtm0_plog_rec_fini(struct m0_dtm0_log_rec **dl_lrec,
+				  struct m0_be_dtm0_log   *log,
+				  struct m0_be_tx         *tx)
+{
+	struct m0_be_seg       *seg = log->dl_seg;
+	struct m0_dtm0_log_rec *rec = *dl_lrec;
+
+	M0_BE_FREE_PTR_SYNC(rec->dlr_txd.dtd_ps.dtp_pa, seg, tx);
+	M0_BE_FREE_PTR_SYNC(rec->dlr_payload.b_addr, seg, tx);
+	M0_BE_FREE_PTR_SYNC(rec, seg, tx);
+	*dl_lrec = NULL;
+}
+
+static int m0_be_dtm0_log__insert(struct m0_be_dtm0_log  *log,
+				  struct m0_be_tx        *tx,
+				  struct m0_dtm0_tx_desc *txd,
+				  struct m0_buf          *payload)
+{
+	int                      rc;
+	struct m0_dtm0_log_rec	*rec;
+	struct m0_be_seg	*seg = log->dl_seg;
+
+	if (log->dl_is_persistent) {
+		rc = be_dtm0_plog_rec_init(&rec, tx, seg, txd, payload);
+		if (rc != 0)
+			return rc;
+		lrec_be_tlink_create(rec, tx);
```

Comment:
Can we put the calling of lrec_be_tlink_create() into the function be_dtm0_plog_rec_init()?

Reply:
This will be reworked as a part of https://jts.seagate.com/browse/EOS-19332 in the next sprint, closing.